### PR TITLE
support more button features 

### PR DESCRIPTION
**Pull request by folkertdev**
This PR adds 

* radio and checkbox elements as buttons (see [the bootstrap docs](https://v4-alpha.getbootstrap.com/components/buttons/#checkbox-and-radio-buttons))
in the Button and ButtonGroup module. 
* `Button.onClick`  for convenience and because radioButton and checkboxButton need preventDefault=True on the "click" event to work properly. 

I've implemented the above features because I needed them. to completely cover the bootstrap buttons we could also add `Button.toggle` which sets `data-toggle="button"`.

### CODE REVIEW

**Comment by rundis** [comment]
Thx @folkertdev . I´m going to look at it next week. Need a little breather after the release.
Some quick questions tho;
- would you typically mix radio and/or checkbox buttons with other buttons in a group ?
- for radios it might make sense to offer a group function that took name and a list of radiobuttons (helps ensure all radios within a group gets the same name)

**Comment by folkertdev** [comment]
Totally understandable, this project is massive. 

* Mixing different types of buttons in the same group is probably very confusing. 
* yes that's a good idea

Both of the above points suggest that more type safety for button groups is a good idea.
This would mean defining four wrapper types with their own constructor, and group and groupItem functions. 

```elm
type RadioButton msg = RadioButton (Html msg)
-- etc for the other buttons

radioButton : Bool -> List (Button.Option msg) -> List (Html.Html msg) -> RadioButton msg

radioButtonGroup : String -> List (RadioButton msg) -> Html msg

radioButtonGroupItem : String -> List (RadioButton msg) -> GroupItem msg
``` 
This is tedious, but type safe.  I'm not sure how safe you want this package to be. The increase in complexity might not be worth the gain in safety.

**Comment by rundis** [comment]
My thinking about type safety so far has been; As type safe as possible while still being nice to use (priority 1) and don't take away to much flexibility (priority 2). Me writing/maintaining less code has been pretty low on the priority scale thus far ( I might come to regret that 😄  ). Bootstrap.Grid.Col is a boilerplate nightmare, but the best I could come up with in terms of being succinct and reasonably nice to use.

So I think your suggestions makes sense. A slight cringe over the extra bool parameter, but can't think of a nice way to avoid it either (other than leaving users to use Button.attrs which isn't too nice in this instance).

**Comment by folkertdev** [comment]
makes sense. I'll implement that then. 

Would it help if I split the `Button.onClick` into a separate PR? that way we can have specific design discussion about just the button here. On the other hand, `Button.onClick fixes an issue that the radio and checkbox buttons introduce, so they are related.

**Comment by rundis** [comment]
nah we can carry on here. np.

**Comment by folkertdev** [comment]
I've quickly written the code for this, to see what it would look like. 

The extra code doesn't look that bad. I was a little worried that adding the types would make the API look more complicated, but it seems to be OK. If you agree, I'll polish it up. 

To what extend should the opaque wrappers and functions have doc comments? I dislike the empty doc comment, but repeating the same message (an opaque type representing a * button) over and over again has little value. thoughts? 

aside: is it normal for commits from the original repo to show up as an added commit here (e.g. your version bump)? did I do something wrong when merging this branch into my fork?

**Comment by rundis** [comment]
Well that's not normal/right with the commits. Did you rebase your branch before updating the PR ?
Regarding comments, yeah I've been struggling with that one, but felt obliged to write something.  

Code looks fine to me, so if you manage to make the pr be one commit containing the changes I think we are good to go for merging.

**Comment by folkertdev** [comment]
I did some polishing. 

All exposed values have a comment (some more helpful than others), and radio- and checkbuttonGroup have examples. 

A note from the last commit message that should be discussed at some point: 

> An unadded feature is the `name` attribute for a group of radio buttons.
> 
> Radio buttons are represented as a label enclosing a radio input
> 
>     <div role="group" data-toggle="buttons" class="btn-group">
>         <label class="active btn btn-secondary" data-toggle="button">
>             <input type="radio" autocomplete="off">On
>         </label>
>         <label class=" btn btn-secondary" data-toggle="button">
>             <input type="radio" autocomplete="off">Off
>         </label>
>     </div>
> 
> The problem is that the bootstrap attributes must be applied to the
> surrounding label. Thus, the Options that `Button.radioButton` gets as
> an argument are for the label, not for the input. Therefore, it is
> impossible to set the name attribute for the `input` element.
> 
> I'm not sure whether this is a problem for elm. In practice, a radio
> button group will be driven by some sum type, for instance
> 
>     type Light = On | Off
> 
> your model will contain this value
> 
>     type alias Model = { light : Light }
> 
> So, regardless of the dom, it is impossible to have 2 or fewer
> than 1 radio button selected, because such a state cannot be expressed.
> 
> So I suggest we don't implement this for now, and see whether we can
> find a use case for "one name for all radios in a group" within an elm
> application.

Apparently the merger can [squash commits from a pull request into one](https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits). That way the individual commits are preserved, but the git history of the master isn't polluted. 

OT: 

* Testing: sure, any areas that need work in particular (besides the new buttons)?
* CI: no experience with this

File: src/Bootstrap/ButtonGroup.elm [hunk 1]
```diff
@@ -57,12 +72,30 @@ type GroupItem msg
...
-{-| Opaque type representing a button or link button, for composing button groups
+{-| Opaque type representing a button, for composing button groups
 -}
 type ButtonItem msg
     = ButtonItem (Html.Html msg)
 
 
+{-| Opaque type representing a link button, for composing button groups
+-}
+type LinkButtonItem msg
```
Comment (rundis) [inline src/Bootstrap/ButtonGroup.elm]: Sorry, but this one passed me by the last time. 

Does this have to be a separate type ? I mean mixing link buttons and ordinary buttons (sorry) shouldn't really be a problem and can't think of a really good reason to not allow that ?

Comment (folkertdev) [inline src/Bootstrap/ButtonGroup.elm]: I was going for consistency. 

I can't really think of cases where I'd mix links and buttons (would seem confusing to me, just as mixing radio and checkbox buttons is), but no good reason to disallow it either.

Comment (rundis) [inline src/Bootstrap/ButtonGroup.elm]: Ah well, let's go with the way it is and see if anyone comes up with a valid use case for changing it !

**Comment by rundis** [comment]
Thank you !

**Pull request merged**